Sources/StreamChat/Repositories/ConnectionRepository.swift

```diff
@@ -0,0 +1,225 @@
+        set { connectionQueue.async(flags: .barrier) { self._connectionStatus = newValue }}
+    }
+
+    /// The current connection id
+    private(set) var connectionId: ConnectionId? {
+        get { connectionQueue.sync { _connectionId } }
+        set { connectionQueue.async(flags: .barrier) { self._connectionId = newValue }}
+    }
+
+    let isClientInActiveMode: Bool
+    private let syncRepository: SyncRepository
+    private let webSocketClient: WebSocketClient?
+    private let apiClient: APIClient
+    private let timerType: Timer.Type
+
+    init(
+        isClientInActiveMode: Bool,
+        syncRepository: SyncRepository,
+        webSocketClient: WebSocketClient?,
+        apiClient: APIClient,
+        timerType: Timer.Type
+    ) {
+        self.isClientInActiveMode = isClientInActiveMode
+        self.syncRepository = syncRepository
+        self.webSocketClient = webSocketClient
+        self.apiClient = apiClient
+        self.timerType = timerType
+    }
+
+    /// Connects the chat client the controller represents to the chat servers.
+    ///
+    /// When the connection is established, `ChatClient` starts receiving chat updates, and `currentUser` variable is available.
+    ///
+    /// - Parameters:
+    ///   - userInfo:       The user information that will be created OR updated if it exists.
+    ///   - completion: Called when the connection is established. If the connection fails, the completion is called with an error.
+    ///
+    func connect(
+        userInfo: UserInfo? = nil,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        // Connecting is not possible in connectionless mode (duh)
+        guard isClientInActiveMode else {
+            completion?(ClientError.ClientIsNotInActiveMode())
+            return
+        }
+
+        guard connectionId == nil else {
+            log.warning("The client is already connected. Skipping the `connect` call.")
+            completion?(nil)
+            return
+        }
+
+        // Set up a waiter for the new connection id to know when the connection process is finished
+        provideConnectionId { [weak webSocketClient] result in
+            switch result {
+            case .success:
+                completion?(nil)
+            case .failure:
+                // Try to get a concrete error
+                if case let .disconnected(source) = webSocketClient?.connectionState {
+                    completion?(ClientError.ConnectionNotSuccessful(with: source.serverError))
+                } else {
+                    completion?(ClientError.ConnectionNotSuccessful())
+                }
+            }
+        }
+        webSocketClient?.connect()
+    }
+
+    /// Disconnects the chat client the controller represents from the chat servers. No further updates from the servers
+    /// are received.
+    func disconnect(
+        source: WebSocketConnectionState.DisconnectionSource = .userInitiated,
```

Comment:
Can we actually make this argument without a default value? On the client side, maybe it makes sense, but here in the repository, it makes sense to not omit this decision 🤔

Reply:
Good point